enzy_htp/structure/structure_constraint/api.py

```diff
@@ -1,43 +1,553 @@
+                target_value:float,
+                params:Dict[str, Any]
+                ):
+        """Ctor that exits if the supplied number of atoms is NOT correct."""
+        
+        self.atoms_ = atoms 
+        self.target_value_ = target_value 
+        self.params_ = deepcopy( params ) 
+
+        self['calc_method'] = params.get('calc_method', 'rosetta')
+
+        if not self.correct_num_atoms():
+            _LOGGER.error("Incorrect number of atoms supplied! Exiting...")
+            exit( 1 )
+
+
+    #TODO(CJ): add function that checks if topology and constraints are compatible
+    #TODO(CJ): will need to make a version of this that actually works for the ResiduePairConstraint
+
+    def clone(self) -> StructureConstraint:
+        """Clones the StructureConstraint with the original target_value."""
+        return self.__init__(atoms, self.target_value, deepcopy(self.params_))
+
+    def clone_current(self) -> StructureConstraint:
+        """Get a version of the StructureConstraint with the current value as the 
+        target value."""
+        return self.__init__(atoms, self.current_value, deepcopy(self.params_))
+
+    def is_cartesian_freeze(self) -> bool:
+        """Is this a cartesian freeze constraint?"""
+        return False
+
+    def is_distance_constraint(self) -> bool:
+        """Is this a distance constraint?"""
+        return False
+
+    def is_angle_constraint(self) -> bool:
+        """Is this an angle constraint?"""
+        return False
+
+    def is_dihedral_constraint(self) -> bool:
+        """Is this a dihedral constraint?"""
+        return False
+
+    def is_residue_pair_constraint(self) -> bool:
+        """Is this a residue pair constraint?"""
+        return False
+
+    @abstractmethod
+    def correct_num_atoms(self) -> bool:
+        """Does the given constraint have the correct number of atoms?"""
+        pass
+
+    def score_energy(self) -> float:
+        """Scores the energy penalty of the constraint. If the calc method is not specified, the method exits."""
+        if self['calc_method'] == 'rosetta':
+            penalty:float = self['penalty'] 
+            tolerance:float = self['tolerance']
+            difference:float = abs(self.current_geometry() - self.target_value)
+
+            if difference <= tolerance:
+                return 0.0
+            else:
+                return penalty * (difference - tolerance)
+
         else:
-            return self._is_backbone_freeze
+            _LOGGER.error(f"Unspecified calculation method {self['calc_method']}. Supported include: rosetta. Exiting...")
+            exit( 1 )
+
+
+    def change_topology(self, new_topology:Structure) -> None:
+        """Switch the associated topology and atoms for a given Constraint. Maps the atom keys
+        as described in Atom.key() and Structure.get_atom().
+
+        Args:
+            new_topology: new structure to translate atoms to.
+
+        Returns:
+            Nothing.
```

Comment:
This seems does nothing lol

Reply:
huh? wym

Reply:
It does not raise an error but just returns a boolean

Reply:
Oh the location of this is also wrong. I will make a new comment.